File: mobile/lib/models/member.dart
import 'enums.dart';

class Member {
  final String id;
  final String parishId;
  final String? familyId;
  final String? sccId;
  final String memberCode;
  final String firstName;
  final String? middleName;
  final String lastName;
  final String? dateOfBirth;
  final GenderType? gender;
  final MaritalStatus? maritalStatus;
  final String? nationalId;
  final String? occupation;
  final String? email;
  final String? phoneNumber;
  final String? physicalAddress;
  final String? photoUrl;
  final bool? isHeadOfFamily;
  final String? notes;
  final bool? isActive;
  final String? createdAt;
  final String? updatedAt;

  Member({
    required this.id,
    required this.parishId,
    this.familyId,
    this.sccId,
    required this.memberCode,
    required this.firstName,
    this.middleName,
    required this.lastName,
    this.dateOfBirth,
    this.gender,
    this.maritalStatus,
    this.nationalId,
    this.occupation,
    this.email,
    this.phoneNumber,
    this.physicalAddress,
    this.photoUrl,
    this.isHeadOfFamily,
    this.notes,
    this.isActive,
    this.createdAt,
    this.updatedAt,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'],
      parishId: json['parish_id'],
      familyId: json['family_id'],
      sccId: json['scc_id'],
      memberCode: json['member_code'],
      firstName: json['first_name'],
      middleName: json['middle_name'],
      lastName: json['last_name'],
      dateOfBirth: json['date_of_birth'],
      gender: json['gender'] != null
          ? enumFromString(GenderType.values, json['gender'])
          : null,
      maritalStatus: json['marital_status'] != null
          ? enumFromString(MaritalStatus.values, json['marital_status'])
          : null,
      nationalId: json['national_id'],
      occupation: json['occupation'],
      email: json['email'],
      phoneNumber: json['phone_number'],
      physicalAddress: json['physical_address'],
      photoUrl: json['photo_url'],
      isHeadOfFamily: json['is_head_of_family'],
      notes: json['notes'],
      isActive: json['is_active'],
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'parish_id': parishId,
      'family_id': familyId,
      'scc_id': sccId,
      'member_code': memberCode,
      'first_name': firstName,
      'middle_name': middleName,
      'last_name': lastName,
      'date_of_birth': dateOfBirth,
      'gender': gender != null ? enumToString(gender!) : null,
      'marital_status': maritalStatus != null ? enumToString(maritalStatus!) : null,
      'national_id': nationalId,
      'occupation': occupation,
      'email': email,
      'phone_number': phoneNumber,
      'physical_address': physicalAddress,
      'photo_url': photoUrl,
      'is_head_of_family': isHeadOfFamily,
      'notes': notes,
      'is_active': isActive,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}

File: backend/src/models/member.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;
use chrono::{NaiveDate, DateTime, Utc};

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, sqlx::Type)]
#[sqlx(type_name = "gender_type", rename_all = "SCREAMING_SNAKE_CASE")]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum GenderType {
    Male,
    Female,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, sqlx::Type)]
#[sqlx(type_name = "marital_status", rename_all = "SCREAMING_SNAKE_CASE")]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum MaritalStatus {
    Single,
    Married,
    Widowed,
    Separated,
    Divorced,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, sqlx::Type)]
#[sqlx(type_name = "sacrament_type", rename_all = "SCREAMING_SNAKE_CASE")]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum SacramentType {
    Baptism,
    FirstCommunion,
    Confirmation,
    Marriage,
    HolyOrders,
    AnointingOfSick,
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, sqlx::Type)]
#[sqlx(type_name = "family_role", rename_all = "SCREAMING_SNAKE_CASE")]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum FamilyRole {
    Head,
    Spouse,
    Member,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct Member {
    pub id: Uuid,
    pub parish_id: Uuid,
    pub family_id: Option<Uuid>,
    pub scc_id: Option<Uuid>,
    pub member_code: String,
    pub first_name: String,
    pub middle_name: Option<String>,
    pub last_name: String,
    pub date_of_birth: Option<NaiveDate>,
    pub gender: Option<GenderType>,
    pub marital_status: Option<MaritalStatus>,
    pub national_id: Option<String>,
    pub occupation: Option<String>,
    pub email: Option<String>,
    pub phone_number: Option<String>,
    pub physical_address: Option<String>,
    pub photo_url: Option<String>,
    pub family_role: Option<FamilyRole>,
    pub notes: Option<String>,
    pub is_active: Option<bool>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub deleted_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct SacramentRecord {
    pub id: Uuid,
    pub member_id: Uuid,
    pub sacrament_type: SacramentType,
    pub sacrament_date: NaiveDate,
    pub officiating_minister: Option<String>,
    pub parish_id: Uuid,
    pub church_name: Option<String>,
    pub certificate_number: Option<String>,
    pub godparent_1_name: Option<String>,
    pub godparent_2_name: Option<String>,
    pub spouse_id: Option<Uuid>,
    pub spouse_name: Option<String>,
    pub witnesses: Option<String>,
    pub notes: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub deleted_at: Option<DateTime<Utc>>,
}

